File: dotnet/ESExpr.Runtime/ESExprBinaryWriter.cs
using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.IO;
using System.Numerics;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace ESExpr.Runtime;

public class ESExprBinaryWriter {
	
	public ESExprBinaryWriter(IImmutableList<string> symbolTable, Stream stream) {
		this.symbolTable = symbolTable;
		this.stream = stream;
	}
	
	private readonly IImmutableList<string> symbolTable;
	private readonly Stream stream;


	public async Task Write(Expr expr, CancellationToken cancellationToken = default) {
		switch(expr) {
			case Expr.Constructor(var constructor, var args, var kwargs):
				switch(constructor) {
					case StringTable.Codec.StringTableConstructor:
						await WriteToken(new BinToken(BinToken.TokenType.ConstructorStartStringTable, null), cancellationToken).ConfigureAwait(false);
						break;
					
					case VList<int>.Codec.ListConstructor:
						await WriteToken(new BinToken(BinToken.TokenType.ConstructorStartList, null), cancellationToken).ConfigureAwait(false);
						break;
						
					default:
						var index = GetSymbolIndex(constructor);
						await WriteToken(new BinToken(BinToken.TokenType.Constructor, index), cancellationToken).ConfigureAwait(false);
						break;
				}

				foreach(var arg in args) {
					await Write(arg, cancellationToken).ConfigureAwait(false);
				}

				foreach(var kvp in kwargs) {
					await WriteToken(new BinToken(BinToken.TokenType.Keyword, GetSymbolIndex(kvp.Key)), cancellationToken).ConfigureAwait(false);
					await Write(kvp.Value, cancellationToken).ConfigureAwait(false);
				}
				
				await WriteToken(new BinToken(BinToken.TokenType.ConstructorEnd, null), cancellationToken).ConfigureAwait(false);
				
				break;
			
			case Expr.Bool(true):
				await WriteToken(new BinToken(BinToken.TokenType.True, null), cancellationToken).ConfigureAwait(false);
				break;
			
			case Expr.Bool(false):
				await WriteToken(new BinToken(BinToken.TokenType.False, null), cancellationToken).ConfigureAwait(false);
				break;
			
			case Expr.Int(var i):
				if(i.Sign >= 0) {
					await WriteToken(new BinToken(BinToken.TokenType.Int, i), cancellationToken).ConfigureAwait(false);
				}
				else {
					await WriteToken(new BinToken(BinToken.TokenType.NegInt, (-i) - 1), cancellationToken).ConfigureAwait(false);
				}
				break;

			case Expr.Str(var s):
			{
				var b = Encoding.UTF8.GetBytes(s);
				await WriteToken(new BinToken(BinToken.TokenType.String, b.Length), cancellationToken).ConfigureAwait(false);
				await stream.WriteAsync(b, cancellationToken).ConfigureAwait(false);
				break;
			}
				
			case Expr.Binary(var b):
				await WriteToken(new BinToken(BinToken.TokenType.Binary, b.Length), cancellationToken).ConfigureAwait(false);
				await stream.WriteAsync(b, cancellationToken).ConfigureAwait(false);
				break;

			case Expr.Float32(var f):
			{
				await WriteToken(new BinToken(BinToken.TokenType.Float32, null), cancellationToken).ConfigureAwait(false);	
				
				uint bits = unchecked((uint)BitConverter.SingleToInt32Bits(f));
				for(int i = 0; i < sizeof(uint); ++i) {
					byte b = unchecked((byte)bits);
					await stream.WriteAsync(new byte[] { b }, cancellationToken).ConfigureAwait(false);
					bits >>= 8;
				}
				
				break;
			}
				
			case Expr.Float64(var d):
			{
				await WriteToken(new BinToken(BinToken.TokenType.Float64, null), cancellationToken).ConfigureAwait(false);	
				
				ulong bits = unchecked((ulong)BitConverter.DoubleToInt64Bits(d));
				for(int i = 0; i < sizeof(ulong); ++i) {
					byte b = unchecked((byte)bits);
					await stream.WriteAsync(new byte[] { b }, cancellationToken).ConfigureAwait(false);
					bits >>= 8;
				}
				
				break;
			}
			
			case Expr.Null:
				await WriteToken(new BinToken(BinToken.TokenType.Null, null), cancellationToken).ConfigureAwait(false);
				break;
				
			default:
				throw new InvalidOperationException();
		}
	}

	private BigInteger GetSymbolIndex(string constructor) {
		int index = symbolTable.IndexOf(constructor);
		if(index < 0) {
			throw new SyntaxException();
		}

		return index;
	}

	private async Task WriteToken(BinToken binToken, CancellationToken cancellationToken = default) {
		byte b = binToken.BinTokenType switch {
			BinToken.TokenType.Constructor => 0x00,
			BinToken.TokenType.Int => 0x20,
			BinToken.TokenType.NegInt => 0x40,
			BinToken.TokenType.String => 0x60,
			BinToken.TokenType.StringPoolIndex => 0x80,
			BinToken.TokenType.Binary => 0xA0,
			BinToken.TokenType.Keyword => 0xC0,
			BinToken.TokenType.ConstructorEnd => 0xE0,
			BinToken.TokenType.True => 0xE1,
			BinToken.TokenType.False => 0xE2,
			BinToken.TokenType.Null => 0xE3,
			BinToken.TokenType.Float32 => 0xE4,
			BinToken.TokenType.Float64 => 0xE5,
			BinToken.TokenType.ConstructorStartStringTable => 0xE6,
			BinToken.TokenType.ConstructorStartList => 0xE7,
			_ => throw new InvalidOperationException(),
		};

		if(binToken.IntValue is { } intValue) {
			b |= (byte)(intValue & 0x0F);
			intValue >>= 4;

			var isPos = intValue.Sign > 0;
			if(isPos) {
				b |= 0x10;
			}

			await stream.WriteAsync(new byte[] { b }, cancellationToken).ConfigureAwait(false);

			if(isPos) {
				await WriteInt(intValue, cancellationToken).ConfigureAwait(false);
			}
		}
		else {
			await stream.WriteAsync(new byte[] { b }, cancellationToken).ConfigureAwait(false);
		}
	}

	private async ValueTask WriteInt(BigInteger intValue, CancellationToken cancellationToken = default) {
		byte[] buff = new byte[1];
		do {
			byte b = (byte)(intValue & 0x7F);
			intValue >>= 7;

			if(intValue.Sign > 0) {
				b |= 0x80;
			}

			buff[0] = b;
			await stream.WriteAsync(buff, cancellationToken).ConfigureAwait(false);
		} while(intValue.Sign > 0);
	}


	public static StringTable BuildSymbolTable(Expr expr) {
		var builder = new SymbolTableBuilder();
		builder.Add(expr);
		return builder.Build();
	}

	public sealed class SymbolTableBuilder {
		private readonly ISet<string> st = new HashSet<string>();
		
		public void Add(Expr expr) {
			if(expr is Expr.Constructor(var name, var args, var kwargs)) {
				if(name != VList<int>.Codec.ListConstructor && name != StringTable.Codec.StringTableConstructor) {
					st.Add(name);
				}

				foreach(var arg in args) {
					Add(arg);
				}
				
				foreach(var kvp in kwargs) {
					st.Add(kvp.Key);
					Add(kvp.Value);
				}
			}
		}

		public StringTable Build() => new StringTable(st.ToImmutableList());
	}
}

File: java/lib/src/main/java/dev/argon/esexpr/ESExprBinaryWriter.java
package dev.argon.esexpr;

import com.google.common.collect.ImmutableList;

import java.io.IOException;
import java.io.OutputStream;
import java.math.BigInteger;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Objects;
import java.util.Set;

/**
 * Encodes ESExpr values into a binary format.
 */
public class ESExprBinaryWriter {

	/**
	 * Creates an encoder.
	 * @param symbolTable The initial symbol table used when writing.
	 * @param os The stream.
	 */
	public ESExprBinaryWriter(List<? extends String> symbolTable, OutputStream os) {
		this.symbolTable = new ArrayList<>(symbolTable);
		symbolSet = new HashSet<>(symbolTable);
		this.os = os;
	}

	/**
	 * Creates an encoder.
	 * @param os The stream.
	 */
	public ESExprBinaryWriter(OutputStream os) {
		symbolTable = new ArrayList<>();
		symbolSet = new HashSet<>();
		this.os = os;
	}

	private final List<String> symbolTable;
	private final Set<String> symbolSet;
	private final OutputStream os;

	/**
	 * Write an ESExpr to the stream.
	 * @param expr The ESExpr to write.
	 * @throws IOException when an error occurs in the underlying stream.
	 */
	public void write(ESExpr expr) throws IOException {
		int oldSize = symbolTable.size();
		addSymbols(expr);
		if(symbolTable.size() > oldSize) {
			writeToken(BinToken.Fixed.APPEND_STRING_TABLE);

			if(oldSize + 1 == symbolTable.size()) {
				writeExprRaw(new ESExpr.Str(symbolTable.get(oldSize)));
			}
			else {
				var appendTable = new StringTable(
					ImmutableList.copyOf(symbolTable.subList(oldSize, symbolTable.size()))
				);
				writeExprRaw(StringTable.codec().encode(appendTable));
			}
		}

		writeExprRaw(expr);
	}

	private void writeExprRaw(ESExpr expr) throws IOException {
		switch(expr) {
			case ESExpr.Constructor(var constructor, var args, var kwargs) -> {
				Objects.requireNonNull(constructor);
				switch(constructor) {
					case BinToken.StringTableName -> writeToken(BinToken.Fixed.CONSTRUCTOR_START_STRING_TABLE);
					case BinToken.ListName -> writeToken(BinToken.Fixed.CONSTRUCTOR_START_LIST);
					case BinToken.MapName -> writeToken(BinToken.Fixed.CONSTRUCTOR_START_MAP);
					case BinToken.SetName -> writeToken(BinToken.Fixed.CONSTRUCTOR_START_SET);
					default -> {
						var index = getSymbolIndex(constructor);
						writeToken(new BinToken.WithInteger(BinToken.WithIntegerType.CONSTRUCTOR, index));
					}
				}
				for(var arg : args) {
					writeExprRaw(arg);
				}
				for(var pair : kwargs.entrySet()) {
					var index = getSymbolIndex(pair.getKey());
					writeToken(new BinToken.WithInteger(BinToken.WithIntegerType.KEYWORD, index));
					writeExprRaw(pair.getValue());
				}
				writeToken(BinToken.Fixed.CONSTRUCTOR_END);
			}

			case ESExpr.Bool(var b) -> {
				if(b) {
					writeToken(BinToken.Fixed.TRUE);
				}
				else {
					writeToken(BinToken.Fixed.FALSE);
				}
			}

			case ESExpr.Int(var i) -> {
				if(i.signum() < 0) {
					writeToken(new BinToken.WithInteger(BinToken.WithIntegerType.NEG_INT, i.negate().subtract(BigInteger.ONE)));
				}
				else {
					writeToken(new BinToken.WithInteger(BinToken.WithIntegerType.INT, i));
				}
			}

			case ESExpr.Str(var s) -> {
				byte[] b = s.getBytes(StandardCharsets.UTF_8);
				writeToken(new BinToken.WithInteger(BinToken.WithIntegerType.STRING, BigInteger.valueOf(b.length)));
				os.write(b);
			}

			case ESExpr.Float16(var f) -> {
				writeToken(BinToken.Fixed.FLOAT16);
				int bits = Short.toUnsignedInt(f);
				for(int i = 0; i < 2; ++i) {
					os.write(bits & 0xFF);
					bits >>>= 8;
				}
			}

			case ESExpr.Float32(var f) -> {
				writeToken(BinToken.Fixed.FLOAT32);
				int bits = Float.floatToRawIntBits(f);
				for(int i = 0; i < 4; ++i) {
					os.write(bits & 0xFF);
					bits >>>= 8;
				}
			}

			case ESExpr.Float64(var d) -> {
				writeToken(BinToken.Fixed.FLOAT64);
				long bits = Double.doubleToRawLongBits(d);
				for(int i = 0; i < 8; ++i) {
					os.write((int)bits & 0xFF);
					bits >>>= 8;
				}
			}

			case ESExpr.Array8(var b) -> {
				writeToken(new BinToken.WithInteger(BinToken.WithIntegerType.ARRAY8, BigInteger.valueOf(b.size())));

				for(int i = 0; i < b.size(); ++i) {
					os.write(b.get(i) & 0xFF);
				}
			}

			case ESExpr.Array16(var b) -> {
				writeToken(BinToken.Fixed.ARRAY16);
				writeInt(BigInteger.valueOf(b.size()));
				for(int i = 0; i < b.size(); ++i) {
					short value = b.get(i);
					os.write(value & 0xFF);
					os.write((value >> 8) & 0xFF);
				}
			}

			case ESExpr.Array32(var b) -> {
				writeToken(BinToken.Fixed.ARRAY32);
				writeInt(BigInteger.valueOf(b.size()));
				for(int i = 0; i < b.size(); ++i) {
					int value = b.get(i);
					os.write(value & 0xFF);
					os.write((value >> 8) & 0xFF);
					os.write((value >> 16) & 0xFF);
					os.write((value >> 24) & 0xFF);
				}
			}

			case ESExpr.Array64(var b) -> {
				writeToken(BinToken.Fixed.ARRAY64);
				writeInt(BigInteger.valueOf(b.size()));
				for(int i = 0; i < b.size(); ++i) {
					long value = b.get(i);
					os.write((int) value & 0xFF);
					os.write((int) (value >> 8) & 0xFF);
					os.write((int) (value >> 16) & 0xFF);
					os.write((int) (value >> 24) & 0xFF);
					os.write((int) (value >> 32) & 0xFF);
					os.write((int) (value >> 40) & 0xFF);
					os.write((int) (value >> 48) & 0xFF);
					os.write((int) (value >> 56) & 0xFF);
				}
			}

			case ESExpr.Array128(var b) -> {
				if((b.size() % 2) != 0) {
					throw new IllegalArgumentException("Array128 must have even length");
				}
				writeToken(BinToken.Fixed.ARRAY128);
				writeInt(BigInteger.valueOf(b.size() / 2));
				for(int i = 0; i < b.size(); ++i) {
					long value = b.get(i);
					os.write((int) value & 0xFF);
					os.write((int) (value >> 8) & 0xFF);
					os.write((int) (value >> 16) & 0xFF);
					os.write((int) (value >> 24) & 0xFF);
					os.write((int) (value >> 32) & 0xFF);
					os.write((int) (value >> 40) & 0xFF);
					os.write((int) (value >> 48) & 0xFF);
					os.write((int) (value >> 56) & 0xFF);
				}
			}

			case ESExpr.Null(var level) -> {
				if(level.equals(BigInteger.ZERO)) {
					writeToken(BinToken.Fixed.NULL0);
				}
				else if(level.equals(BigInteger.ONE)) {
					writeToken(BinToken.Fixed.NULL1);
				}
				else if(level.equals(BigInteger.valueOf(2))) {
					writeToken(BinToken.Fixed.NULL2);
				}
				else {
					writeToken(BinToken.Fixed.NULLN);
					writeInt(level.subtract(BigInteger.valueOf(3)));
				}
			}
		}
	}

	private void writeToken(BinToken token) throws IOException {
		switch(token) {
			case BinToken.WithInteger(var type, var value) -> {
				int b = switch(type) {
					case CONSTRUCTOR -> 0x00;
					case INT -> 0x20;
					case NEG_INT -> 0x40;
					case STRING -> 0x60;
					case STRING_POOL_INDEX -> 0x80;
					case ARRAY8 -> 0xA0;
					case KEYWORD -> 0xC0;
				};

				b |= value.byteValue() & 0x0F;
				value = value.shiftRight(4);

				boolean isPos = value.signum() > 0;
				if(isPos) {
					b |= 0x10;
				}
				os.write(b);
				if(isPos) {
					writeInt(value);
				}
			}
			case BinToken.Fixed fixed -> {
				int b = switch(fixed) {
					case CONSTRUCTOR_END -> 0xE0;
					case TRUE -> 0xE1;
					case FALSE -> 0xE2;
					case NULL0 -> 0xE3;
					case NULL1 -> 0xE8;
					case NULL2 -> 0xE9;
					case NULLN -> 0xEA;
					case FLOAT16 -> 0xEC;
					case FLOAT32 -> 0xE4;
					case FLOAT64 -> 0xE5;
					case CONSTRUCTOR_START_STRING_TABLE -> 0xE6;
					case CONSTRUCTOR_START_LIST -> 0xE7;
					case CONSTRUCTOR_START_MAP -> 0xF1;
					case CONSTRUCTOR_START_SET -> 0xF2;
					case APPEND_STRING_TABLE -> 0xEB;
					case ARRAY16 -> 0xED;
					case ARRAY32 -> 0xEE;
					case ARRAY64 -> 0xEF;
					case ARRAY128 -> 0xF0;
				};
				os.write(b);
			}
		}
	}

	private BigInteger getSymbolIndex(String symbol) throws IOException {
		int index = symbolTable.indexOf(symbol);
		if(index < 0) {
			index = symbolTable.size();
			writeToken(BinToken.Fixed.APPEND_STRING_TABLE);
			writeExprRaw(new ESExpr.Str(symbol));
		}
		return BigInteger.valueOf(index);
	}

	private void writeInt(BigInteger value) throws IOException {
		do {
			int b = value.byteValue() & 0x7F;
			value = value.shiftRight(7);

			if(value.signum() > 0) {
				b |= 0x80;
			}
			os.write(b);
		} while(value.signum() > 0);
	}


	private void addSymbols(ESExpr expr) {
		if(expr instanceof ESExpr.Constructor(var name, var args, var kwargs)) {
			if(!name.equals(BinToken.StringTableName) && !name.equals(BinToken.ListName)) {
				addSymbol(name);
			}

			for(var arg : args) {
				addSymbols(arg);
			}

			for(var kwarg : kwargs.entrySet()) {
				addSymbol(kwarg.getKey());
				addSymbols(kwarg.getValue());
			}
		}
	}

	private void addSymbol(String symbol) {
		if(symbolSet.add(symbol)) {
			symbolTable.add(symbol);
		}
	}

}
